From: Werner Koch <wk@gnupg.org>
Date: Tue, 29 Nov 2016 20:15:47 +0100
Subject: agent,dirmngr: Handle corner case in shutdown mode.

* agent/gpg-agent.c (handle_connections): Keep on selecting on the
inotify fd even when a shutdown is pending.
* dirmngr/dirmngr.c (handle_connections): Ditto.  Also simplifyy the
use of the HAVE_INOTIFY_INIT cpp conditional.
--

Without that patch we won't notice a removed socket when a shutdown is
pending.  This is somewhat related to bug report 2849.

Signed-off-by: Werner Koch <wk@gnupg.org>
(cherry picked from commit 854adc8ae19749e44cb79dfa0c5401f48012b13a)
---
 agent/gpg-agent.c |  6 ++++++
 dirmngr/dirmngr.c | 12 +++++++-----
 2 files changed, 13 insertions(+), 5 deletions(-)

diff --git a/agent/gpg-agent.c b/agent/gpg-agent.c
index fa67a3a..79d58bd 100644
--- a/agent/gpg-agent.c
+++ b/agent/gpg-agent.c
@@ -2872,6 +2872,12 @@ handle_connections (gnupg_fd_t listen_fd,
           /* Do not accept new connections but keep on running the
              loop to cope with the timer events.  */
           FD_ZERO (&fdset);
+          nfd = -1;
+          if (my_inotify_fd != -1)
+            {
+              FD_SET (my_inotify_fd, &fdset);
+              nfd = my_inotify_fd;
+            }
 	}
 
       /* POSIX says that fd_set should be implemented as a structure,
diff --git a/dirmngr/dirmngr.c b/dirmngr/dirmngr.c
index e03aa33..21500ec 100644
--- a/dirmngr/dirmngr.c
+++ b/dirmngr/dirmngr.c
@@ -1839,9 +1839,7 @@ handle_connections (assuan_fd_t listen_fd)
   int nfd, ret;
   fd_set fdset, read_fdset;
   int saved_errno;
-#ifdef HAVE_INOTIFY_INIT
-  int my_inotify_fd;
-#endif /*HAVE_INOTIFY_INIT*/
+  int my_inotify_fd = -1;
 
   npth_attr_init (&tattr);
   npth_attr_setdetachstate (&tattr, NPTH_CREATE_DETACHED);
@@ -1883,14 +1881,12 @@ handle_connections (assuan_fd_t listen_fd)
   FD_ZERO (&fdset);
   FD_SET (FD2INT (listen_fd), &fdset);
   nfd = FD2INT (listen_fd);
-#ifdef HAVE_INOTIFY_INIT
   if (my_inotify_fd != -1)
     {
       FD_SET (my_inotify_fd, &fdset);
       if (my_inotify_fd > nfd)
         nfd = my_inotify_fd;
     }
-#endif /*HAVE_INOTIFY_INIT*/
 
   /* Main loop.  */
   for (;;)
@@ -1904,6 +1900,12 @@ handle_connections (assuan_fd_t listen_fd)
           /* Do not accept new connections but keep on running the
              select loop to wait for signals (e.g. SIGCHLD).  */
           FD_ZERO (&fdset);
+          nfd = -1;
+          if (my_inotify_fd != -1)
+            {
+              FD_SET (my_inotify_fd, &fdset);
+              nfd = my_inotify_fd;
+            }
 	}
 
       /* Take a copy of the fdset.  */
